moveit_core/trajectory_processing/src/ruckig_traj_smoothing.cpp

```diff
@@ -0,0 +1,226 @@
+    // If the requested velocity is too great, a joint can actually "move backward" to give itself more time to
+    // accelerate to the target velocity. Iterate and decrease velocities until that behavior is gone.
+    bool backward_motion_detected = checkForBackwardMotion(num_dof, ruckig_input, ruckig_output);
+
+    double minimum_velocity_magnitude = 0.01;  // rad/s
+    double velocity_magnitude = getTargetVelocityMagnitude(ruckig_input, num_dof);
+    while (backward_motion_detected && rclcpp::ok() && (velocity_magnitude > minimum_velocity_magnitude))
+    {
+      // decrease target velocity
+      for (size_t joint = 0; joint < num_dof; ++joint)
+      {
+        ruckig_input.target_velocity[joint] *= 0.9;
+      }
+      velocity_magnitude = getTargetVelocityMagnitude(ruckig_input, num_dof);
+      // Run Ruckig
+      ruckig_result = ruckig.update(ruckig_input, ruckig_output);
+      // check for backward motion
+      backward_motion_detected = checkForBackwardMotion(num_dof, ruckig_input, ruckig_output);
+    }
+
+    if (ruckig_result != ruckig::Result::Working)
+    {
+      RCLCPP_ERROR_STREAM(LOGGER, "Ruckig trajectory smoothing failed at waypoint " << waypoint_idx);
+      RCLCPP_ERROR_STREAM(LOGGER, "Ruckig error: " << ruckig_result);
+      return false;
+    }
+
+    // Overwrite pos/vel/accel of the target waypoint
+    for (size_t joint = 0; joint < num_dof; ++joint)
+    {
+      next_waypoint->setVariablePosition(idx[joint], ruckig_output.new_position[joint]);
+      next_waypoint->setVariableVelocity(idx[joint], ruckig_output.new_velocity[joint]);
+      next_waypoint->setVariableAcceleration(idx[joint], ruckig_output.new_acceleration[joint]);
+    }
+  }
+
+  return true;
+}
+
+double RuckigSmoothing::getTargetVelocityMagnitude(const ruckig::InputParameter<0>& ruckig_input, const size_t num_dof)
+{
+  double vel_magnitude = 0;
+  for (size_t joint = 0; joint < num_dof; ++joint)
+  {
+    vel_magnitude += ruckig_input.target_velocity[joint] * ruckig_input.target_velocity[joint];
+  }
+  return sqrt(vel_magnitude);
+}
+
+bool RuckigSmoothing::checkForBackwardMotion(const size_t num_dof, const ruckig::InputParameter<0>& ruckig_input,
+                                             const ruckig::OutputParameter<0>& ruckig_output)
+{
+  // Check for backward motion of any joint
+  for (size_t joint = 0; joint < num_dof; ++joint)
+  {
+    // Negative result indicates different signs --> opposing directions of motion
+    if (((ruckig_output.new_position[joint] - ruckig_input.current_position[joint]) /
+         ruckig_input.target_velocity[joint]) < 1)
+    {
+      return true;
+      break;
+    }
+  }
+  return false;
+}
+
+void RuckigSmoothing::getNextCurrentTargetStates(ruckig::InputParameter<0>& ruckig_input,
+                                                 ruckig::OutputParameter<0>& ruckig_output,
+                                                 const moveit::core::RobotStatePtr next_waypoint, const size_t num_dof,
```

Comment:
Const shared pointers can force copies because to actually share the pointer you need to be able to update the reference count.  Use either a const& or a non-const shared ptr.  Also, get rid of the const on the size_t